### PR TITLE
retire: Performance adjustment & Minor naming tweak

### PR DESCRIPTION
**Pull request by kingthorin**
- CHANGELOG > Added change note.
- Messages.properties > Updated name.
- Repo.java > Added script element extraction for html responses.
- RetireScanRule > Pass Jericho Source object instead of re-creating it.

Fixes zaproxy/zaproxy#6959

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

### CODE REVIEW

File: addOns/retire/CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,9 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
-
+### Changed
+- Performance improvements (Issue 6959).
```
Comment (thc202) [inline addOns/retire/CHANGELOG.md:8]: Did you check that this improves? (It should but would be good to confirm with a file with similar size.)

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java [hunk 2]
```diff
@@ -128,6 +131,19 @@ public Result scanJS(HttpMessage msg) {
         return scanHash(hash);
     }
 
+    private String getCleanContent(HttpMessage msg) {
+        if (msg.getResponseHeader().isHtml()) {
+            Source source = new Source(msg.getResponseBody().toString());
```
Comment (thc202) [inline addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java:136]: Better use existing `Source`.

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java [hunk 3]
```diff
@@ -128,6 +131,19 @@ public Result scanJS(HttpMessage msg) {
...
+            StringBuilder contents = new StringBuilder();
+            for (Element scriptElement : source.getAllElements(HTMLElementName.SCRIPT)) {
+                contents.append(scriptElement.toString());
+                contents.append('\n');
+            }
+            return contents.toString();
+        } else {
```
Comment (thc202) [inline addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java:143]: Remove the else.

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java (hunk 2 above)
Comment (kingthorin) [reply addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java:136]: The method is "public" can I still change the signature? (Just do a full version bump in the add-on?)

File: addOns/retire/CHANGELOG.md (hunk 1 above)
Comment (kingthorin) [reply addOns/retire/CHANGELOG.md:8]: Yes. I'll do some more live testing and find a message to include in unit tests I guess?

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java (hunk 2 above)
Comment (thc202) [reply addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java:136]: Who's calling this class directly? If you really want to keep binary compatibility leave the old method (which creates the `Source`) calling the new one.

File: addOns/retire/CHANGELOG.md (hunk 1 above)
Comment (thc202) [reply addOns/retire/CHANGELOG.md:8]: I don't think it's necessary to add a real file, a dummy with random content should replicate too.

Comment (kingthorin) [reply addOns/retire/CHANGELOG.md:8]: I can't find a good one to save but here's my example:

```bash
Original

159972 [ZAP-PassiveScan-3] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 5 seconds to scan https://cdn.syndication.twimg.com/timeline/profile?callback=__twttr.callbacks.tl_i0_profile_HoCChamber_old&dnt=false&domain=www.ourcommons.ca&lang=en&screen_name=HoCChamber&suppress_response_codes=true&t=1838258&tz=GMT-0400&with_replies=false application/javascript;charset=utf-8 129242
160045 [ZAP-PassiveScan-3] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 5 seconds to scan https://cdn.syndication.twimg.com/timeline/profile?callback=__twttr.callbacks.tl_i0_profile_HoCChamber_old&dnt=false&domain=www.ourcommons.ca&lang=en&screen_name=HoCChamber&suppress_response_codes=true&t=1838258&tz=GMT-0400&with_replies=false application/javascript;charset=utf-8 129242
166568 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 15 seconds to scan https://platform.twitter.com/widgets/widget_iframe.d7fc2fc075c61f6fa34d79a0cbbf1e34.html?origin=https%3A%2F%2Fwww.ourcommons.ca text/html; charset=utf-8 327158
166694 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 9 seconds to scan https://www.parl.ca/legisinfo/en/overview text/html; charset=utf-8 67545
166726 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 9 seconds to scan https://www.parl.ca/legisinfo/en/overview text/html; charset=utf-8 67545
166756 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 16 seconds to scan https://platform.twitter.com/widgets/widget_iframe.d7fc2fc075c61f6fa34d79a0cbbf1e34.html?origin=https%3A%2F%2Fwww.ourcommons.ca text/html; charset=utf-8 327158
170400 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 28 seconds to scan https://www.ourcommons.ca/en text/html; charset=utf-8 289228
170610 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 28 seconds to scan https://www.ourcommons.ca/en text/html; charset=utf-8 289228
171939 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 5 seconds to scan https://www.parl.ca/CommonElementsV3/js/jquery-3.5.1.min.js application/javascript 89478
172002 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 5 seconds to scan https://www.parl.ca/CommonElementsV3/js/jquery-3.5.1.min.js application/javascript 89478
181226 [ZAP-PassiveScan-3] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 9 seconds to scan https://petitions.ourcommons.ca/en/Home/Index text/html; charset=utf-8 74747
181273 [ZAP-PassiveScan-3] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 9 seconds to scan https://petitions.ourcommons.ca/en/Home/Index text/html; charset=utf-8 74747
185080 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 8 seconds to scan https://petitions.ourcommons.ca/assets/site.css?v=2 text/css 268881
185083 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 8 seconds to scan https://petitions.ourcommons.ca/assets/site.css?v=2 text/css 268881
192056 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 14 seconds to scan https://www.gstatic.com/recaptcha/releases/M-QqaF9xk6BpjLH22uHZRhXt/recaptcha__en.js text/javascript 370206
192188 [ZAP-PassiveScan-4] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 14 seconds to scan https://www.gstatic.com/recaptcha/releases/M-QqaF9xk6BpjLH22uHZRhXt/recaptcha__en.js text/javascript 370206

195534 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 12 seconds to scan https://www.gstatic.com/recaptcha/releases/M-QqaF9xk6BpjLH22uHZRhXt/recaptcha__en.js text/javascript 370206
195663 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library took 12 seconds to scan https://www.gstatic.com/recaptcha/releases/M-QqaF9xk6BpjLH22uHZRhXt/recaptcha__en.js text/javascript 370206

251193 [AWT-EventQueue-0] INFO  org.zaproxy.zap.extension.autoupdate.ExtensionAutoUpdate - Installing new addon retire v0.13.0
251197 [AWT-EventQueue-0] INFO  org.zaproxy.zap.extension.pscan.ExtensionPassiveScan - loaded passive scan rule: Vulnerable JS Library (Retire.js)
251200 [AWT-EventQueue-0] INFO  org.zaproxy.zap.extension.autoupdate.ExtensionAutoUpdate - Finished installing new addon retire v0.13.0

282437 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 6 seconds to scan https://petitions.ourcommons.ca/assets/site.css?v=2 text/css 268881
282439 [ZAP-PassiveScan-1] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library (Retire.js) took 6 seconds to scan https://petitions.ourcommons.ca/assets/site.css?v=2 text/css 268881
376124 [AWT-EventQueue-0] INFO  org.parosproxy.paros.control.Control - Discard Session
376169 [AWT-EventQueue-0] INFO  org.parosproxy.paros.control.Control - New Session
376170 [AWT-EventQueue-0] INFO  org.parosproxy.paros.control.Control - Create and Open Untitled Db
376191 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache commit start
376276 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache commit end
376332 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - Database closed
376507 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache open start
376632 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache commit start
376716 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache commit end
376716 [AWT-EventQueue-0] INFO  hsqldb.db.HSQLDB379AF3DEBD.ENGINE - dataFileCache open end

New

398937 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Wappalyzer Scanner (Tech Detection) took 11 seconds to scan https://platform.twitter.com/widgets/widget_iframe.d7fc2fc075c61f6fa34d79a0cbbf1e34.html?origin=https%3A%2F%2Fwww.ourcommons.ca text/html; charset=utf-8 327158
399036 [ZAP-PassiveScan-2] WARN  org.zaproxy.zap.extension.pscan.PassiveScanTask - Passive Scan rule Vulnerable JS Library (Retire.js) took 11 seconds to scan https://platform.twitter.com/widgets/widget_iframe.d7fc2fc075c61f6fa34d79a0cbbf1e34.html?origin=https%3A%2F%2Fwww.ourcommons.ca text/html; charset=utf-8 327158
```

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java (hunk 2 above)
Comment (kingthorin) [reply addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java]: I doubt anyone else is, but I was trying to do things "right".

File: addOns/retire/CHANGELOG.md (hunk 1 above)
Comment (thc202) [reply addOns/retire/CHANGELOG.md:8]: Great, no need to add a unit test IMO.

File: addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java (hunk 2 above)
Comment (thc202) [reply addOns/retire/src/main/java/org/zaproxy/addon/retire/model/Repo.java]: This is more an internal class though, which should not be part of the public API.

File: addOns/retire/CHANGELOG.md (hunk 1 above)
Comment (kingthorin) [reply addOns/retire/CHANGELOG.md:8]: Okay, I'll look into it. So far 20MB of lorem ipsum doesn't slow it down (this or the original), might actually need to be HTML.

Comment (kingthorin) [reply addOns/retire/CHANGELOG.md:8]: > Great, no need to add a unit test IMO.

Thanks

**Comment by kingthorin** [comment]
Will tweak the commit message in a bit the gradle.properties note is no longer needed.

**Comment by kingthorin** [comment]
Tweaked

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by ricekot — APPROVED** [review]
Thank you

File: addOns/retire/src/main/resources/org/zaproxy/addon/retire/resources/Messages.properties [hunk 4]
```diff
@@ -1,6 +1,6 @@
 retire.desc = Retire.js
 
-retire.rule.name = Vulnerable JS Library
+retire.rule.name = Vulnerable JS Library (Retire.js)
```
[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/retire/src/main/resources/org/zaproxy/addon/retire/resources/Messages.properties:3]: Maybe `Vulnerable JS Library (via Retire.js)` or `Vulnerable JS Library (by Retire.js)` would be better?

[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/retire/src/main/resources/org/zaproxy/addon/retire/resources/Messages.properties:3]: Thanks

[the remark this answers is not in the document]
Comment (kingthorin) [reply addOns/retire/src/main/resources/org/zaproxy/addon/retire/resources/Messages.properties:3]: We decided to separate the values. The Rule name will now include "(Powered by Retire.js)" but the Alerts will remain as they previously were.

**Pull request merged**